### DISCUSSION TITLE
`asyncpg`: getting the original exception from DBAPIError

### DISCUSSION THREAD

**Discussion by ljiatu** [sqlalchemy/sqlalchemy#6199] [Usage Questions] [answered]
First of all, thank you so much for the hard work on getting asyncio support released! It's so great to be able to get rid of all the stuff we had for adapting the sync API to an async web server framework.

My question is about the proper way to get the original exception from `DBAPIError` when using the `asyncpg` adapter. When using `psycopg2`, we were able to inspect the `pgcode` property on the `orig` exception to see what type of error got thrown, but I'm not seeing something equivalent for the `asyncpg ` sdapter. After reading through the exception handling code, it seems like `_handle_exception` in `asyncpg.py` is translating the exceptions raised by `asyncpg` to an `Error` class that doesn't appear to have accessor methods to access the original exception. Rather, it puts the type of the original exception and the message into a string. So is string parsing the best way to get the original exception for now? We are interested in the original error since we are retrying on serialization errors.

Thanks!

**Comment by zzzeek** [comment, marked as answer]
OK if we can find that on asyncpg's exceptions we can add that.

Here's a sample error I made:

```
        await conn.execute(
            t1.insert(), [{"id": 1, "name": "some name 1"}, {"id": 1, "name": "some name 2"}]
        )

```

then we get this trace:

```
sqlalchemy.exc.IntegrityError: (sqlalchemy.dialects.postgresql.asyncpg.IntegrityError) <class 'asyncpg.exceptions.UniqueViolationError'>: duplicate key value violates unique constraint "t1_pkey"
DETAIL:  Key (id)=(1) already exists.
[SQL: INSERT INTO t1 (id, name) VALUES (%s, %s)]
[parameters: ((1, 'some name 1'), (1, 'some name 2'))]

```

if I pdb in and look at asyncpg's object, I see the following.  is it "sqlstate" you're looking for ?

```
(Pdb) error
UniqueViolationError('duplicate key value violates unique constraint "t1_pkey"')
(Pdb) error.__dict__
{'query': 'INSERT INTO t1 (id, name) VALUES ($1::integer, $2::varchar)', 'severity': 'ERROR', 'severity_en': 'ERROR', 'sqlstate': '23505', 'message': 'duplicate key value violates unique constraint "t1_pkey"', 'detail': 'Key (id)=(1) already exists.', 'schema_name': 'public', 'table_name': 't1', 'constraint_name': 't1_pkey', 'server_source_filename': 'nbtinsert.c', 'server_source_line': '563', 'server_source_function': '_bt_check_unique'}

```

**Reply by ljiatu** [reply]
Either `sqlstate` or the original exception class itself (i.e. `asyncpg.exceptions.UniqueViolationError`) works!

**Reply by zzzeek** [reply]
asyncpg's exceptions aren't in a pep-249 hierarchy in any case and SQLAlchemy's adapter should try to make things behave as closely to psycpopg2 as possible,  so we can maybe add the accessor "sqlstate" and mirror it as "pgcode".

**Reply by ljiatu** [reply]
That works. Thanks.

**Reply by zzzeek** [reply]
that's merged in d01790ebe0b425bf79c4decf24011d0d6234944d

**Reply by ljiatu** [reply]
Thank you!!

**Reply by theo-brown** [reply]
@zzzeek it would also be useful to be able to extract the `'message'` and `'detail'` properties of asyncpg's object, so  that when the error is raised you can inform the user where the conflict is  - is there any way of doing that?

**Reply by theo-brown** [reply]
For example, it would make sense if `sqlalchemy.exc.IntegrityError.detail` is set to `f"{UniqueViolationError.message}: {UniqueViolationError.detail}"`

**Reply by zzzeek** [reply]
.detail on SQLAlhcemy exception is a list.  To get at the "original" exception from the DBAPI, you call upon the ".orig" attribute; that's where you'd get additional elements that are specific to asyncpg's exception class.

however, as we have an additional level of proxying for asyncpg, the change to SQLAlchemy would be to make sure the sqlalchemy.dialects.postgresql.asyncpg.dbapi.IntegrityError class has the appropriate asyncpg elements transferred over to it.  We can accept PRs for this.

**Reply by zzzeek** [reply]
see the commit mentioned above at d01790ebe0b425bf for how to transfer additional asyncpg fields to the ".orig" attribute including tests.

**Reply by theo-brown** [reply]
Cool, I'll have a go at a PR imitating d01790e

**Reply by martyanov** [reply]
@zzzeek so there is  still no way to get the original `asyncpg` exception and ideally re-raise it? Maybe I miss something, but if it's true, then async variant of `sqlalchemy` is mostly not usable in this regard at the moment, as it's even not possible to get the violated constraint name.

**Reply by martyanov** [reply]
Ok, probably we can use `error.orig.__cause__`  and `error.orig.__context__`, reconstruct the original exception and get all the missed informaton, but it's neither intuitive, nor friendly. Again, maybe I miss something?

**Reply by zzzeek** [reply]
please open a bug report as we dont track issues mentioned in discussions.

I think the original exception object is preserved in `__cause__` in any case as we do re-raise it appropriately.  Please open a bug report so that the issue can be tracked.  We will also be asking for PRs welcome on this one in order to expedite.

**Reply by amacfie-tc** [reply]
Issue created https://github.com/sqlalchemy/sqlalchemy/issues/9843
